File: control.swift
public var g1 = Grid()
public var g2 = Grid()
public var g3 = Grid()

public struct control{  

    var dimensions : Int
    var gridInit : GridInit
    var tfsf: TotalFieldScatteringField
    var abc : AbsorbingBoundaryCondition
    var snapshot : Snapshot
    var update : UpdateFunction
    var sourceFunction : SourceFunction

    public mutating func initialize(){
        
        if self.dimensions == 1{
            g1 = self.gridInit.initialize()
            self.tfsf.initialize(g : g1)
            self.abc.initialize(g : g1)
            self.snapshot.initialize(g : g1)
            
        }
        
        else if self.dimensions == 2{
            g2 = self.gridInit.initialize()
            self.tfsf.initialize(g : g2)
            self.abc.initialize(g : g2)
            self.snapshot.initialize(g : g2)
            
        }
        if self.dimensions == 3{
            g3 = self.gridInit.initialize()
            self.tfsf.initialize(g : g3)
            self.abc.initialize(g : g3)
            self.snapshot.initialize(g : g3)
        }
        
    }

    public mutating func iterate(maxTime: Int){

        for t in 0...maxTime{
            
            if self.dimensions == 1{
                g1.time = t
                self.update.magneticField(g : g1)
                self.abc.apply(g : g1)
                self.update.electricField(g : g1)
                self.tfsf.update(g : g1)
                self.sourceFunction.apply(time : Double(g1.time), location : 0.0)
                self.snapshot.snap(g : g1)
            }

            else if self.dimensions == 2{
                g2.time = t
                self.update.magneticField(g : g2)
                self.abc.apply(g : g2)
                self.update.electricField(g : g2)
                self.tfsf.update(g : g2)
                self.sourceFunction.apply(time : Double(g2.time), location : 0.0)
                self.snapshot.snap(g : g2)
            }

            else{
                g3.time = t
                self.update.magneticField(g : g3)
                self.abc.apply(g : g3)
                self.update.electricField(g : g3)
                self.tfsf.update(g : g3)
                self.sourceFunction.apply(time : Double(g3.time), location : 0.0)
                self.snapshot.snap(g : g3)
            }
           
        }
        
        writeToFile(FileURL: "dataFile", dataToWrite: DataString)
    }
}

public func create_remote() -> control{
    
    var gridInit : GridInit
    var tfsf: TotalFieldScatteringField
    var abc : AbsorbingBoundaryCondition
    var snapshot = Snapshot()
    var update : UpdateFunction
    var sourceFunction : SourceFunction
    var type : Int

    print("Enter dimensions of grid")
    let dimensions = intInput()
    if dimensions == 1{
        
        tfsf = plain1Dtfsf() as TotalFieldScatteringField
        
        print("1 - absorbing boundary condition, 0 - reflection")
        let boundaryType = intInput()

        if boundaryType == 1{
            abc = plain1Dabc() as AbsorbingBoundaryCondition
        }

        else{
            abc = emptyAbc() as AbsorbingBoundaryCondition
        }
        
        gridInit = plain1DGridInit() as GridInit
        update = oneDUpdate() as UpdateFunction
        sourceFunction = ezInc() as SourceFunction
    }

    else if dimensions == 2 {

        print("1 - tmz, 2 - tez")
        let type2D = intInput()

        if type2D == 1{
            
            type = 1
            tfsf = tmztfsf() as TotalFieldScatteringField
            gridInit = tmzTfsfGridInit() as GridInit
            update = tmzUpdate() as UpdateFunction
            abc = tmzabc() as AbsorbingBoundaryCondition
        }

        else{
            type = 2
            tfsf = teztfsf() as TotalFieldScatteringField
            gridInit = tezGridInit() as GridInit
            update = tezUpdate() as UpdateFunction
            abc = tezabc() as AbsorbingBoundaryCondition
        }

        print("1 - absorbing boundary condition, 0 - reflection")
        let boundaryType = intInput()

        if boundaryType == 0{
            abc = emptyAbc() as AbsorbingBoundaryCondition
        }
        
        sourceFunction = Ricker2D() as SourceFunction
    }
    else{

        print("1 - normal, 0 - sphere")
        let type3D = intInput()
        

        tfsf = tfsf3d() as TotalFieldScatteringField
        update = update3D() as UpdateFunction
        abc = abc3d() as AbsorbingBoundaryCondition
        
        if type3D == 1{
            gridInit = GridInit3D(isSpherePresent : false) as GridInit
        }
        else{
            gridInit = GridInit3D(isSpherePresent : true) as GridInit
        }
        
        print("1 - absorbing boundary condition, 0 - reflection")
        let boundaryType = intInput()

        if boundaryType == 0{
            abc =  emptyAbc() as AbsorbingBoundaryCondition
        }
        
        sourceFunction = Ricker2D() as SourceFunction
    }
    
    print("1 - additive source 0 - tfsf? ")
    let sourceInitiator = intInput()
    
    if sourceInitiator == 1{
        tfsf = emptytfsf() as TotalFieldScatteringField
        sourceFunction = additiveSourceFunction() as SourceFunction
    }


    return control(dimensions:dimensions, gridInit : gridInit, tfsf: tfsf, abc : abc, snapshot : snapshot, update : update, sourceFunction : sourceFunction)
}

public func runSimulation(){

    var ben = create_remote()
    ben.initialize()
    ben.iterate(maxTime : 450)
}

File: control.py
from grid import *
from gridInit import *
from tfsf import *
from ezinc import *
from update import *
from snapshot import *
from abc import *

g1 = Grid
g2 = Grid
g3 = Grid

class control:  

    
    def __init__(self,dimensions, gridInit, tfsf, abc, snapshot, update ,sourceFunction):
        self.dimensions = dimensions
        self.gridInit = gridInit
        self.tfsf = tfsf
        self.abc = abc
        self.snapshot = snapshot
        self.update = update
        self.sourceFunction = sourceFunction
    
    def initialize(self):
        
        if self.dimensions == 1:
            g1 = self.gridInit.initialize()
            self.tfsf.initialize(g1)
            self.abc.initialize(g1)
            self.snapshot.initialize(g1)
            
        
        
        elif self.dimensions == 2:
            g2 = self.gridInit.initialize()
            self.tfsf.initialize(g2)
            self.abc.initialize(g2)
            self.snapshot.initialize(g2)
            
        
        if self.dimensions == 3:
            g3 = self.gridInit.initialize()
            self.tfsf.initialize(g3)
            self.abc.initialize(g3)
            self.snapshot.initialize(g3)
        
        
    

    def iterate(maxTime):

        for t in range(maxTime):
            
            if self.dimensions == 1:
                g1.time = t
                self.update.magneticField(g1)
                self.abc.apply(g1)
                self.update.electricField(g1)
                self.tfsf.update(g1)
                self.sourceFunction.apply(float(g1.time),  0.0)
                self.snapshot.snap(g1)
            

            elif self.dimensions == 2:
                g2.time = t
                self.update.magneticField(g2)
                self.abc.apply(g2)
                self.update.electricField(g2)
                self.tfsf.update(g2)
                self.sourceFunction.apply( float(g2.time),  0.0)
                self.snapshot.snap(g2)
            

            else:
                g3.time = t
                self.update.magneticField(g3)
                self.abc.apply(g3)
                self.update.electricField(g3)
                self.tfsf.update(g3)
                self.sourceFunction.apply( float(g3.time),  0.0)
                self.snapshot.snap(g3)
            
            print(ez1)
        

def create_remote() :
    
    gridInit = GridInit()
    tfsf = TotalFieldScatteringField()
    abc = AbsorbingBoundaryCondition()
    snapshot = Snapshot()
    update = UpdateFunction()
    sourceFunction =  SourceFunction()
    Type = 0 

    print("Enter dimensions of grid")
    dimensions = intInput()
    if dimensions == 1:
        
        tfsf = plain1Dtfsf
        
        print("1 - absorbing boundary condition, 0 - reflection")
        boundaryType = intInput()

        if boundaryType == 1:
            abc = plain1Dabc()
        

        else:
            abc = emptyAbc()
        
        
        gridInit = plain1DGridInit()
        update = oneDUpdate()
        sourceFunction = ezInc()
    

    elif dimensions == 2 :

        print("1 - tmz, 2 - tez")
        type2D = intInput()

        if type2D == 1:
            
            Type = 1
            tfsf = tmztfsf()
            gridInit = tmzTfsfGridInit()
            update = tmzUpdate()
            abc = tmzabc()
        

        else:
            Type = 2
            tfsf = teztfsf()
            gridInit = tezGridInit()
            update = tezUpdate()
            abc = tezabc()
        

        print("1 - absorbing boundary condition, 0 - reflection")
        boundaryType = intInput()

        if boundaryType == 0:
            abc = emptyAbc()
        
        
        sourceFunction = Ricker2D()
    
    else:

        print("1 - normal, 0 - sphere")
        type3D = intInput()
        

        tfsf = tfsf3d()
        update = update3D()
        abc = abc3d()
        
        if type3D == 1:
            gridInit = GridInit3D(false) 
        
        else:
            gridInit = GridInit3D(true) 
        
        
        print("1 - absorbing boundary condition, 0 - reflection")
        boundaryType = intInput()

        if boundaryType == 0:
            abc =  emptyAbc()
        
        
        sourceFunction = Ricker2D()
    
    
    print("1 - additive source 0 - tfsf? ")
    sourceInitiator = intInput()
    
    if sourceInitiator == 1:
        tfsf = emptytfsf()
        sourceFunction = additiveSourceFunction()
    


    return control(dimensions,  gridInit,  tfsf,  abc, snapshot,  update,  sourceFunction)


def runSimulation():

    ben = create_remote()
    ben.initialize()
    ben.iterate(450)


runSimulation()